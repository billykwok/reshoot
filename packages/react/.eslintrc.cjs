module.exports = {
  root: false,
  env: {
    browser: true,
    node: true,
    commonjs: true,
    es6: true,
    'jest/globals': true,
  },
  parser: '@babel/eslint-parser',
  parserOptions: {
    ecmaVersion: 'latest',
    sourceType: 'module',
    ecmaFeatures: { impliedStrict: true, jsx: true },
    babelOptions: { configFile: './babel.config.cjs' },
  },
  settings: {
    react: { version: 'detect' },
    'import/extensions': ['.js', '.jsx', '.json'],
    'import/resolver': { node: { moduleDirectory: ['node_modules'] } },
  },
  plugins: [
    'import',
    'prettier',
    'react',
    'react-hooks',
    'jsx-a11y',
    'testing-library',
    'jest',
  ],
  extends: [
    'eslint:recommended',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    'plugin:prettier/recommended',
    'prettier',
    'plugin:testing-library/react',
    'plugin:jest/recommended',
  ],
  rules: {
    'react/prop-types': 'off',
    'react/jsx-uses-react': 'off',
    'react/react-in-jsx-scope': 'off',
    'jsx-a11y/image-has-content': 'off',
    'jsx-a11y/heading-has-content': 'off',
    'jsx-a11y/anchor-has-content': 'off',
    'react-hooks/rules-of-hooks': 'error',
    'prefer-spread': 'off',
    'consistent-return': 'off',
    'max-len': 'off',
    'no-console': 'off',
  },
  overrides: [
    {
      files: ['*.ts', '*.tsx'],
      parser: '@typescript-eslint/parser',
      parserOptions: {
        ecmaVersion: 'latest',
        sourceType: 'module',
        ecmaFeatures: { impliedStrict: true, jsx: true },
        project: ['./tsconfig.json'],
      },
      settings: {
        react: { version: 'detect' },
        'import/extensions': ['.ts', '.tsx'],
        'import/resolver': { node: { moduleDirectory: ['node_modules'] } },
        'import/parsers': { '@typescript-eslint/parser': ['.ts', '.tsx'] },
      },
      plugins: [
        '@typescript-eslint',
        'import',
        'prettier',
        'react',
        'react-hooks',
        'jsx-a11y',
        'testing-library',
        'jest',
      ],
      extends: [
        'eslint:recommended',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:@typescript-eslint/recommended-requiring-type-checking',
        'plugin:react/recommended',
        'plugin:jsx-a11y/recommended',
        'plugin:prettier/recommended',
        'prettier',
        'plugin:testing-library/react',
        'plugin:jest/recommended',
      ],
      rules: {
        '@typescript-eslint/no-var-requires': 'off',
        '@typescript-eslint/no-explicit-any': 'off',
        '@typescript-eslint/no-floating-promises': 'off',
        '@typescript-eslint/explicit-function-return-type': 'off',
        '@typescript-eslint/explicit-module-boundary-types': [
          'error',
          { allowArgumentsExplicitlyTypedAsAny: true },
        ],
        '@typescript-eslint/prefer-includes': 'off',
        'react/prop-types': 'off',
        'react/jsx-uses-react': 'off',
        'react/react-in-jsx-scope': 'off',
        'jsx-a11y/image-has-content': 'off',
        'jsx-a11y/heading-has-content': 'off',
        'jsx-a11y/anchor-has-content': 'off',
        'react-hooks/rules-of-hooks': 'error',
      },
    },
  ],
};
