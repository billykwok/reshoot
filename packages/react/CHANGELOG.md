# @reshoot/react

## 2.0.5

### Patch Changes

- a0288f4: fix: support hash character in inline options

## 2.0.4

### Patch Changes

- ec76336: fix: improve rollup setup and update dependencies
- Updated dependencies [ec76336]
  - @reshoot/types@2.0.3

## 2.0.3

### Patch Changes

- 0e54ec5: fix: fix style exports

## 2.0.2

### Patch Changes

- 510a7da: fix: fix incorrect extensions in package.json

## 2.0.1

### Patch Changes

- 226a185: fix: add node version restriction and fix rollup config
- Updated dependencies [226a185]
  - @reshoot/types@2.0.1

## 2.0.0

### Major Changes

- 8d162cb: Introducing Reshoot 2

  Reshoot has been rewritten from ground up to support more use cases. Please refer to README.md for details.

### Patch Changes

- Updated dependencies [8d162cb]
  - @reshoot/types@2.0.0
