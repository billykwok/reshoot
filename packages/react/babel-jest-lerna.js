module.exports = require('babel-jest').createTransformer({
  rootMode: 'upward',
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    [
      '@babel/preset-typescript',
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        onlyRemoveTypeImports: true,
      },
    ],
    ['@babel/preset-react', { runtime: 'automatic', useBuiltIns: true }],
    'linaria/babel',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: { version: 3, proposals: true },
        helpers: true,
        regenerator: true,
      },
    ],
    'dynamic-import-node',
    ['const-enum', { transform: 'constObject' }],
  ],
});
