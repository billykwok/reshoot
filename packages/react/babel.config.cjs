module.exports = {
  presets: [
    ['@babel/preset-env', { targets: { esmodules: true }, bugfixes: true }],
    ['@babel/preset-typescript', { isTSX: true, allExtensions: true }],
    ['@babel/preset-react', { runtime: 'automatic', useBuiltIns: true }],
    '@linaria',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      { corejs: { version: 3, proposals: true }, helpers: true },
    ],
  ],
};
