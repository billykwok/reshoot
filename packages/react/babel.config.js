module.exports = {
  presets: [
    ['@babel/preset-env', { targets: { esmodules: true }, bugfixes: true }],
    [
      '@babel/preset-typescript',
      {
        isTSX: true,
        allExtensions: true,
        allowNamespaces: true,
        allowDeclareFields: true,
        onlyRemoveTypeImports: true,
      },
    ],
    '@babel/preset-react',
    'linaria/babel',
  ],
  plugins: [
    [
      '@babel/plugin-transform-runtime',
      {
        corejs: { version: 3, proposals: true },
        helpers: true,
        regenerator: true,
      },
    ],
    ['const-enum', { transform: 'constObject' }],
  ],
};
