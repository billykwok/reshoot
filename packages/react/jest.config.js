import base from '../../jest.config.js';

const config = {
  ...base,
  testEnvironment: 'jsdom',
};

export default config;
