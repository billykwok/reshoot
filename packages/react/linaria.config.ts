export default {
  evaluate: true,
  displayName: process.env.NODE_ENV === 'development',
};
