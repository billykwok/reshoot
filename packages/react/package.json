{
  "name": "@reshoot/react",
  "description": "A React component that lazily renders responsive images with LQIPs",
  "version": "1.6.4",
  "type": "commonjs",
  "exports": {
    "import": "./lib/es/index.js",
    "require": "./lib/index.js"
  },
  "main": "./lib/index.js",
  "types": "./types/index.d.ts",
  "files": [
    "lib",
    "types"
  ],
  "sideEffects": [
    "*.css"
  ],
  "publishConfig": {
    "access": "public"
  },
  "author": "Billy Kwok <opensource@billykwok.me>",
  "license": "MIT",
  "keywords": [
    "reshoot",
    "srcset",
    "image",
    "img",
    "react",
    "responsive",
    "babel"
  ],
  "repository": "https://github.com/billykwok/reshoot",
  "bugs": {
    "url": "https://github.com/billykwok/reshoot/issues"
  },
  "homepage": "https://github.com/billykwok/reshoot/tree/master/packages/react",
  "scripts": {
    "lint": "eslint ./src/**",
    "clean": "rm -rf lib types",
    "build": "yarn clean && rollup -c rollup.config.js && rollup -c rollup.config.js --format es --dir ./lib/es && tsc --project tsconfig.json"
  },
  "peerDependencies": {
    "@babel/runtime-corejs3": "^7.11.2",
    "object-assign": "4.1.1",
    "react": "^16.13.1"
  }
}
