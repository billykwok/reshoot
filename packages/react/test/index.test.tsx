import {
  afterEach,
  beforeAll,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { createRef } from 'react';
import {
  act,
  screen,
  render,
  fireEvent,
  cleanup,
} from '@testing-library/react';
import { type ImageMeta } from '@reshoot/types';

type UseLifecycleRef = typeof import('use-lifecycle-ref').default;
type UseLazyImage = typeof import('../src/useLazyImage').default;

describe('Reshoot', () => {
  const completelyLoadedImgElement = Object.defineProperty(
    document.createElement('img'),
    'complete',
    { value: true }
  );
  const loadingImgElement = Object.defineProperty(
    document.createElement('img'),
    'complete',
    { value: false }
  );
  const imageMeta: ImageMeta = {
    src: 'image.jpg',
    width: 100,
    height: 80,
    aspectRatio: 0.8,
    color: '#fff',
    placeholder: 'data:jpeg;base64,1shfk241jkw',
  };
  const srcSet = 'image-480.jpg 480w, image-640.jpg 640w';
  const sizes = '';
  const alt = 'alt content';
  const className = 'classname-1 classname-2';
  const style = {};
  const useLifecycleRef = jest.fn<UseLifecycleRef>();
  const useLazyImage = jest.fn<UseLazyImage>();

  beforeAll(() => {
    jest.doMock('use-lifecycle-ref', () => useLifecycleRef);
    jest.doMock('../src/useLazyImage', () => useLazyImage);
  });

  afterEach(() => {
    cleanup();
    jest.resetAllMocks();
  });

  test('render input correctly on load with HTMLImageElement.complete true', async () => {
    useLifecycleRef.mockReturnValue(null);
    const { default: Img } = await import('../src');
    const ref = createRef<HTMLElement>();
    const imgRef = createRef<HTMLImageElement>();
    const element = (
      <Img
        meta={imageMeta}
        container="a"
        ref={ref}
        imgRef={imgRef}
        srcSet={srcSet}
        sizes={sizes}
        alt={alt}
        className={className}
        style={style}
        data-testid="123"
      />
    );

    render(element);
    expect(useLazyImage).toHaveBeenCalledTimes(1);
    const [, , , onLoad] = useLazyImage.mock.calls[0];
    expect(useLifecycleRef).toHaveBeenCalledTimes(1);
    const [{ onAttach, ref: outerImgRef }] = useLifecycleRef.mock.calls[0];
    expect(outerImgRef).toEqual(imgRef);
    const dom1 = screen.getByTestId('123');
    expect(dom1).toEqual(ref.current);
    expect(dom1).toMatchSnapshot();

    act(() => {
      onAttach(completelyLoadedImgElement);
    });
    const dom2 = screen.getByTestId('123');
    expect(dom2).toEqual(ref.current);
    expect(dom2).toMatchSnapshot();

    act(() => {
      onLoad(new Event('load'));
    });
    const dom3 = screen.getByTestId('123');
    expect(dom3).toEqual(ref.current);
    expect(dom3).toMatchSnapshot();
  });

  test('render input correctly on load with HTMLImageElement.complete false', async () => {
    useLifecycleRef.mockReturnValue(null);
    const { default: Img } = await import('../src');
    const ref = createRef<HTMLElement>();
    const imgRef = createRef<HTMLImageElement>();
    const element = (
      <Img
        meta={imageMeta}
        container="a"
        ref={ref}
        imgRef={imgRef}
        srcSet={srcSet}
        sizes={sizes}
        alt={alt}
        className={className}
        style={style}
        data-testid="123"
      />
    );

    render(element);
    expect(useLazyImage).toHaveBeenCalledTimes(1);
    const [, , , onLoad] = useLazyImage.mock.calls[0];
    expect(useLifecycleRef).toHaveBeenCalledTimes(1);
    const [{ onAttach, ref: outerImgRef }] = useLifecycleRef.mock.calls[0];
    expect(outerImgRef).toEqual(imgRef);
    const dom1 = screen.getByTestId('123');
    expect(dom1).toEqual(ref.current);
    expect(dom1).toMatchSnapshot();

    act(() => {
      onAttach(loadingImgElement);
    });
    const dom2 = screen.getByTestId('123');
    expect(dom2).toEqual(ref.current);
    expect(dom2).toMatchSnapshot();

    act(() => {
      onLoad(new Event('load'));
    });
    const dom3 = screen.getByTestId('123');
    expect(dom3).toEqual(ref.current);
    expect(dom3).toMatchSnapshot();

    // eslint-disable-next-line testing-library/no-node-access
    const placeholder = dom3.querySelector('[aria-hidden="true"]');
    fireEvent.animationEnd(placeholder);
    expect(dom3).toEqual(ref.current);
    expect(dom3).toMatchSnapshot();
  });

  test('render input correctly on error', async () => {
    useLifecycleRef.mockReturnValue(null);
    const { default: Img } = await import('../src');
    const ref = createRef<HTMLElement>();
    const imgRef = createRef<HTMLImageElement>();
    const element = (
      <Img
        meta={imageMeta}
        ref={ref}
        imgRef={imgRef}
        srcSet={srcSet}
        sizes={sizes}
        alt={alt}
        style={style}
        data-testid="123"
      />
    );

    render(element);
    expect(useLazyImage).toHaveBeenCalledTimes(1);
    const [, , , , onError] = useLazyImage.mock.calls[0];
    expect(useLifecycleRef).toHaveBeenCalledTimes(1);
    const [{ onAttach, ref: outerImgRef }] = useLifecycleRef.mock.calls[0];
    expect(outerImgRef).toEqual(imgRef);
    const dom1 = screen.getByTestId('123');
    expect(dom1).toEqual(ref.current);
    expect(dom1).toMatchSnapshot();

    act(() => {
      onAttach(loadingImgElement);
    });
    const dom2 = screen.getByTestId('123');
    expect(dom2).toEqual(ref.current);
    expect(dom2).toMatchSnapshot();

    act(() => {
      onError(new Event('error'));
    });
    const dom3 = screen.getByTestId('123');
    expect(dom3).toEqual(ref.current);
    expect(dom3).toMatchSnapshot();
  });
});
