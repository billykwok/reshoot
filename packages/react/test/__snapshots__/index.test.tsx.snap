// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reshoot render input correctly on error 1`] = `
<div
  class="cew27ew"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="true"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
</div>
`;

exports[`Reshoot render input correctly on error 2`] = `
<div
  class="cew27ew"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="true"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
</div>
`;

exports[`Reshoot render input correctly on error 3`] = `
<div
  class="cew27ew"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="false"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
  <div
    class="e1b2ugnh"
  >
    <p>
      Failed to load image
    </p>
    <p>
      Please try refreshing the page
    </p>
  </div>
</div>
`;

exports[`Reshoot render input correctly on load with HTMLImageElement.complete false 1`] = `
<a
  class="cew27ew classname-1 classname-2"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="true"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
</a>
`;

exports[`Reshoot render input correctly on load with HTMLImageElement.complete false 2`] = `
<a
  class="cew27ew classname-1 classname-2"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="true"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
</a>
`;

exports[`Reshoot render input correctly on load with HTMLImageElement.complete false 3`] = `
<a
  class="cew27ew classname-1 classname-2"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="true"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only fhsfp3r"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
</a>
`;

exports[`Reshoot render input correctly on load with HTMLImageElement.complete false 4`] = `
<a
  class="cew27ew classname-1 classname-2"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="false"
    crossorigin=""
    src="image.jpg"
  />
</a>
`;

exports[`Reshoot render input correctly on load with HTMLImageElement.complete true 1`] = `
<a
  class="cew27ew classname-1 classname-2"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="true"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
</a>
`;

exports[`Reshoot render input correctly on load with HTMLImageElement.complete true 2`] = `
<a
  class="cew27ew classname-1 classname-2"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="true"
    crossorigin=""
    src="image.jpg"
  />
  <div
    aria-hidden="true"
    class="js-only"
  >
    <img
      alt=""
      src="data:jpeg;base64,1shfk241jkw"
    />
  </div>
</a>
`;

exports[`Reshoot render input correctly on load with HTMLImageElement.complete true 3`] = `
<a
  class="cew27ew classname-1 classname-2"
  data-testid="123"
  sizes=""
  srcset="image-480.jpg 480w, image-640.jpg 640w"
  style="--c: #fff; --r: 0.8%;"
>
  <img
    alt="alt content"
    aria-busy="false"
    crossorigin=""
    src="image.jpg"
  />
</a>
`;
